File: LeetCode146.java
class Node {
    int key;
    int val;
    Node next;
    Node prev;

    public Node(){}

    public Node(int key, int val){
        this.key = key;
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}
class DlinkedList{ // tail is least recently used cache
    Node head;
    Node tail;

    public DlinkedList(){
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.prev = head;
    }
    public void remove(Node n){
        n.prev.next = n.next;
        n.next.prev = n.prev;
    }
    public void inserthead(Node n){
        n.prev = head;
        n.next = head.next;
        head.next.prev = n;
        head.next = n;
    }
    public int removeTail(){
        int key = tail.prev.key;
        remove(tail.prev);
        return key;
    }
}
class LRUCache {
    Map<Integer, Node> cache;
    DlinkedList list;
    int capacity;
    public LRUCache(int capacity) {
        this.capacity = capacity;
        cache = new HashMap();
        list = new DlinkedList();
    }

    public int get(int key) {
        if(!cache.containsKey(key))
            return -1;
        Node n = cache.get(key);
        list.remove(n);
        list.inserthead(n);
        //cache.put(key, n);
        return n.val;
    }

    public void put(int key, int value) {
        Node n = new Node(key, value);
        if(cache.containsKey(key)){
            list.remove(cache.get(key));
        }
        else if(cache.size() >= capacity){
            int k = list.removeTail();
            cache.remove(k);
        }
        list.inserthead(n);
        cache.put(key, n);
    }
}

File: python/LeetCode146.py
class ListNode:
    def __init__(self, val = 0, prev = None, next = None, key = None):
        self.val = val
        self.prev = prev
        self.next = next
        self.key = key
class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.head = ListNode()
        self.tail = ListNode()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.length = 0
        self.cache = {}

    def get(self, key: int) -> int:
        if self.cache.get(key) == None:
            return -1
        else:
            node = self.cache.get(key)
            node.prev.next = node.next
            node.next.prev = node.prev
            node.next = self.head.next
            node.next.prev = node
            self.head.next = node
            node.prev = self.head
            return node.val

    def put(self, key: int, value: int) -> None:
        if self.length == self.capacity and self.cache.get(key) == None:
            node = self.tail.prev
            node.prev.next = self.tail
            self.tail.prev = node.prev
            self.cache.pop(node.key)
            self.length -= 1

        if self.cache.get(key) != None:
            node = self.cache.get(key)
            node.next.prev = node.prev
            node.prev.next = node.next
            node.prev = self.head
            node.next = self.head.next
            node.next.prev = node
            self.head.next = node
            node.val = value
        else:
            node = ListNode()
            node.val = value
            node.key = key
            node.prev = self.head
            node.next = self.head.next
            node.next.prev = node
            self.head.next = node
            self.cache[key] = node
            self.length += 1

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
